software/spmd/pc_profiler/profiler.S

```diff
@@ -0,0 +1,232 @@
+/* 
+  Description:
+  A software-based PC profiler which builds a histogram of PCs encountered by the program
+
+  Memory layout:
+  DRAM --> 16MB (configurable) for maintaing a histogram for N tiles. 
+
+  Registers used:
+  The program being profiled is free to use any registers within if there are no other imposed constraints.
+  The profiler uses the following registers - 
+  t0 --> Base address of the histogram array (actual value stored in dmem[1])
+  t1 --> Scratchpad (actual value stored in dmem[0])
+*/
+
+#include "bsg_manycore_arch.h"
+#include "bsg_manycore_asm.h"
+
+// Useful to control the amount of space required to simulate the use of the profiler.
+// We want 16MB when running on actual hardware.
+#ifndef HIST_SPACE
+#define HIST_SPACE 16777216
+#endif
+
+#ifndef PROFILE
+#define PROFILE 0
+#endif
+
+#if PROFILE > 0
+/* 
+  Data section:
+  Defines a 2 element (8 byte) array to store the base pointer of the histogram array and for use by the interrupt handler to spill some registers
+  dmem[0] = histogram base pointer, temporary storage
+  dmem[1] = temporary storage
+*/
+.section .dmem.interrupt, "aw"
+  _interrupt_arr: .space 8, 0x0
+
+/*
+  DRAM section:
+  Defines a 16MB (array size is configurable) chunk in DRAM to store the histogram.
+
+  Lingering question(s):
+  1. Assigned space per PC (i.e bin capacity) as 32-bits. Is this enough, too much or too less?
+  2. Base pointer in DRAM (for now it is mapped to start of DRAM) [Consider static assignment in linker]
+*/
+.section .dram, "aw"
+.globl _histogram_arr
+  _histogram_arr: .space HIST_SPACE, 0x0
+#endif
+
+/*
+  Startup:
+  This is very similar to the code in software/spmd/common/crt.S and adds only the interrupt service routine (ISR) and sets up the DMEM with the histogram base address. 
+*/
+.section .crtbegin
+.globl _start
+
+#if PROFILE > 0
+/*
+  Remote interrupt handler (should map to address 0)
+*/
+_remote_interrupt:
+  j _remote_interrupt_handler
+
+/*
+  Trace interrupt handler (should map to address 4)
+  Clears the pending trace interrupt bit in the MIP CSR and atomically increments the histogram bin count based on the MEPC CSR value
+
+  The bin index in the histogram can be found using the following equation index = base ptr + EPC
+  (Although EPC is 4-byte aligned, we are assigning 4 bytes per bin, so we don't need to do EPC >> 2)
+
+  Lingering concerns:
+  Since we are using the MEPC, we don't have the PC of the first instruction in the program. 
```

Comment:
Maybe provide more detail about this? I see the instruction where you enable instructions, and then you jump to the code to trace. Presumably the jump instruction is not profiled, but the instructions after the jump are?

Reply:
Yeah, this comment might probably be stale but might still have relevance for smaller programs. With the O2 optimization enabled, smaller functions are actually inlined which means there aren't going to be any jumps in the code. This would result in the very first instruction in the program to not be profiled. 

Two somewhat related question however are - 
1. How small does the function need to be for it to be inlined and what are the typical program sizes for real workloads in the manycore?
2. Even though the compiler inlined the function itself, there still seems to be a dedicated portion of the disassembly that duplicates the function code which increases the code size. Is this something we should care about or am I missing some compiler option here?

Reply:
It's not just how small a function is, but how many times it's used. A medium sized function can be inlined if it's static and only used once.

Which function? If you do not mark the function as `static` the compiler has to generate the actual function even if it inlines it. This is so the function can be called from another source file.

Reply:
The function I'm talking about is the fibonacci function in `main.c`. There is only a call to it from `main()`. Thanks for the explanation though, that clears stuff up for me.